File: PriorityQueue/Medium/Sort_Characters_By_Frequency.java
// Leetcode: https://leetcode.com/problems/sort-characters-by-frequency/

class Solution 
{
        public String frequencySort(String s) 
        {
                HashMap<Character,Integer> map = new HashMap<Character,Integer>();
                for( int i = 0; i < s.length(); i++ )
                {
                        char key = s.charAt(i);
                        if(map.containsKey(key))
                        {
                                map.put(key, map.get(key) + 1);
                        }
                        else
                        {
                                map.put(key, 1);
                        }
                }

                PriorityQueue<Character> pq = new PriorityQueue<Character>((a,b) -> (map.get(b) - map.get(a)));
                for (Character c : map.keySet()) 
                {
                        pq.add(c);
                }

                StringBuilder sb = new StringBuilder();

                while(!pq.isEmpty())
                {
                        char ch = pq.poll();
                        int count = map.get(ch);
                        while(count-- > 0)
                        {
                                sb.append(ch);
                        }
                }
                return sb.toString();
        }
}

File: Array/Medium/Sort_Characters_By_Frequency.js
// Leetcode: https://leetcode.com/problems/sort-characters-by-frequency/

// Solution 1
var frequencySort = (str) => {
    let map = {};
    for (let char of str) {
        map[char] = (map[char] || 0) + 1;
    }

    const sortedChars = Object.entries(map).sort((a, b) => (a[1] === b[1] ? a[0].localeCompare(b[0]) : b[1] - a[1]));

    let result = "";
    for (let char of sortedChars) {
        result += char[0].repeat(char[1]);
    }

    return result;
};

// Solution 2
var frequencySort = (str) => {
    let map = {};
    for (let char of str) {
        map[char] = (map[char] || 0) + 1;
    }

    const sortedChars = Object.entries(map).sort((a, b) => (a[1] === b[1] ? a[0].localeCompare(b[0]) : b[1] - a[1]));

    return sortedChars.reduce((result, char) => result + char[0].repeat(char[1]), "");
};

// Solution 3
const frequencySort = (str) => {
    let map = {};
    for (let char of str) {
        map[char] = (map[char] || 0) + 1;
    }

    let sortedChars = Object.entries(map).sort((a, b) => {
        if (a[1] === b[1]) {
            return a[0].localeCompare(b[0]);
        }
        return b[1] - a[1];
    });

    let result = "";
    for (let char of sortedChars) {
        result += char[0].repeat(char[1]);
    }
    return result;
};

// Solution 4
const frequencySort = (str) => {
    let map = new Map();
    for (let char of str) {
        map.set(char, (map.get(char) || 0) + 1);
    }

    let sortedChars = Array.from(map).sort((a, b) => {
        if (a[1] === b[1]) {
            return a[0].localeCompare(b[0]);
        }
        return b[1] - a[1];
    });

    let result = "";
    for (let char of sortedChars) {
        result += char[0].repeat(char[1]);
    }
    return result;
};

// Solution 5
var frequencySort = (str) => {
    let frequencyMap = {};
    for (let char of str) {
        frequencyMap[char] = (frequencyMap[char] || 0) + 1;
    }

    let sortedChars = Object.keys(frequencyMap).sort((a, b) => {
        if (frequencyMap[a] === frequencyMap[b]) {
            return a.localeCompare(b);
        }
        return frequencyMap[b] - frequencyMap[a];
    });

    let result = "";
    for (let char of sortedChars) {
        result += char.repeat(frequencyMap[char]);
    }

    return result;
};
